dvc/repo/experiments/show.py

```diff
@@ -91,8 +111,9 @@ def show(
     sha_only=False,
     num=1,
     param_deps=False,
+    onerror: Optional[Callable] = None,
```

Comment:
Is there a reason why we are defining our error handler at the command level instead of doing it within `repo.experiments.show`/`repo.metrics.show`?

At least for the show methods (not sure about diff yet), it seems to me like our API would be cleaner if we handled the errors with the `repo...show()` methods rather than the CLI commands. Since we already return a dictionary with revs as keys why not just collect the errors inside the `show()` methods and when some rev fails, just set that revision to something like

```py
{
  "workspace": {...},  # normal rev
  "abc123": {"errors": ["failed parsing foo.yaml"]},
  "def456": {...},  # normal rev
  ...
}
```

At the CLI level, we can collect revs from that dict that have non-empty `errors` and log them as needed with the `Command failed for files in the following revs...` message.

This would also be a straightforward way of passing error messages into VScode, since `--show-json` just dumps the returned dictionary from `metrics.show()`/`experiments.show()`

Reply:
This approach is possible, however I see some things to consider before:
1. In this case we would change how API works, before there was never `errors` in the results.
2. We would need to design how to convert Errors into JSON, as `msg` string is not enough for VScode team, which would like to handle some kinds of errors. So the error jsons would have to contain some information.
In some cases the whole revisions will fail (malformed `dvc.yaml`) and in other particular path (malformed metric file). How do we differentiate between those and let VScode handle the problems at most granular level possible?
Do we include `path,revision` fields in error json objects? Or should we expect that result json can contain the error at any level, like for  example:
```
{
  "workspace": {"path1": {"error":{"message":"File not found", "code": 404}}}, 
  "abc123": {"error": {"message":"failed parsing dvc.yaml", "code":123}},
  "def456": {...},  # normal rev
  ...
}
```

I would personally be leaning towards possible `error` field on any level. What do you think @pmrowla?
@rogermparent @mattseddon would this approach be easier/better for you to handle than `--with-errors` we've been talking about?

Reply:
Regarding 1, we are already going to have to change the API in some way to pass errors to VSCode. I also don't think this is a problem because for anyone who was actually consuming `--show-json` output before, the commands were unusable entirely if an error occurred (since DVC would throw an exception and not return any json output at all).

On 2, what level of granularity we want to use for passing errors probably depends on the specific command/call. For `exp show` I don't think it makes sense to do anything more granular then per experiment rev. But for metrics/params it may be better to do it per file.

Reply:
I think it's okay to change the internal API if it's easier that way. 

Though I'd still prefer that we do not return errors in the `--show-json` without the user explicitly asking for it. We should try to serve more users than the VSCode, and generally, the users don't care about the errors. For VSCode, it could use a private API or a different one easily.

Reply:
I'm not sure whether users want the errors in json. If they are parsing json, they might have similar needs to vscode (unless we end up with special error messages for vscode). Having both options might be good.

It might be just as important or more important to consider what would be best for a public API. Again, having both options seems nice as long as it's easy for users. It might even be good to default to `onerror = OnError` and make is so users can just pass `onerror = None` to exclude them (rather than have to import and pass `OnError` to get the errors).

Reply:
> @pared: We would need to design how to convert Errors into JSON, as `msg` string is not enough for VScode team, which would like to handle some kinds of errors. So the error jsons would have to contain some information.

I think this is a consideration not just for VSCode, but any application consuming these errors programmatically. Many DVC error messages point to specific files or ids, such that the full error cannot completely be converted to something like an error code without some accompanying context. We _could_ parse these from strings, but specifically for machine-reading that seems like bad DX all around.

> would this approach be easier/better for you to handle than `--with-errors` we've been talking about?

I think both approaches shake out to be roughly even in difficulty, since we already have to preprocess all experiments/checkpoints  to get aggregate data like columns.

My personal opinion is that consistently using a separate error field containing error objects containing a unique code and all necessary context would be the most clean, flexible, and low-maintenance way to handle machine-readable errors- I would imagine something like that will be needed anyway for errors that aren't tied to a specific piece of data, so my preference is to be lazily consistent and use that necessary API for all errors instead of having two levels of errors that are consumed slightly differently.

Reply:
> rather than have to import and pass `OnError` to get the errors

Users shouldn't have to import DVC's `OnError` class at all. What we should be documenting is just the callback format (what happens in `OnError.__call__`) so that users can handle errors however they want.

Reply:
The important thing is that however errors end up being communicated in `--show-json`, it should be easy for users to get the same output through the API.

Reply:
> consistently using a separate error field

I agree that having errors separate from the successful results seems best. If I use the API to retrieve experiment results and visualize them, having to handle errors at every possible level first seems painful.